WordPress/gutenberg
# Experimental: allow parent Block to consume child Block's toolbar

Several recent use cases have suggested that it would be valuable to a parent Block rendering child blocks into `InnerBlocks` to be able to ~~"consume"~~ capture the toolbars of the children (in reverse: the children have their toolbars "proxied" to the parent toolbar area to be displayed).

The key benefits of this are:

1. the toolbar remains fixed to the top left avoiding the "toolbar overload" problem associated with having toolbars "follow" the cursor around as the child blocks are selected.
2. the toolbar is rendered in a DOM node divorced from the child Block which works around difficult `overflow: hidden` problems.
3. nested blocks are easier to use because the toolbar is always in a consistent position and not effected by the inner block layout.

~Two specific use cases for this might be:~

~1. Nav Block - see https://github.com/WordPress/gutenberg/pull/18336~
~2. Social Block~

**Update**: for use case please see [Joen's comment below](https://github.com/WordPress/gutenberg/pull/18440#issuecomment-565391080).

In general however, the pattern also lends itself well to any deeply nested Block hierarchies. For example, on the current `master` branch, try nesting a lot of Columns and see how you get on with nested toolbars. It isn't a ideal experience. 

By deferring the _rendering_ of all toolbars to the root block of the hierarchy we provide a better experience.

The PR is experimental and achieves the effect by:

* Rendering a `Slot` for ChildToolbars in all _parents_ of the _currently selected_ Block.
* Conditionally rendering the `BlockContextualToolbar` into the ChildToolbars Slot.

The reason for using Slot/Fill is due to needing to capture events triggered on the childtoolbar within the React component tree rather than the DOM tree. If we allow for the events to bubble in the DOM it triggers selection of the underlying parent Block (which is where the Slot is rendered) which breaks a lot of toolbar functionality.

This works. There are some quirks that would need to be worked out.

I'm looking for feedback about:

1. Whether this goes too far in breaking the existing UI patterns (see Table Block which does something similar _UI wise_ - although it doesn't use `InnerBlocks` or a Parent -> Child relationship)
2. What a11y concerns there might be about having the child toolbar divorced from the block itself and could these be mitigated.
3. Where my code is not considering edge cases.

## How has this been tested?
Manual testing for now.

## Testing

1. Add Nav Menu
2. Add some items
3. Add some nested items
4. Click between the Nav items and see all the child toolbars proxied to the top level Nav Menu.
5. Test keyboard shortcut for jumping to closest toolbar works (ie: `⌥ + F10`)

Also try [switching out the boolean props on `InnerBlocks`](https://github.com/WordPress/gutenberg/blob/18e6909dcbdafaa99916020215494a1388e3b184/packages/block-library/src/navigation-menu/edit.js#L243) on the Nav Menu Block and verify that the documented behaviour works as described.

Also try applying these props to the `InnerBlocks` of the Columns Block. Try nesting!

## Screenshots 
![Screen Capture on 2019-11-15 at 14-52-08](https://user-images.githubusercontent.com/444434/68952561-0061ae80-07b8-11ea-950d-15e9808aa4d9.gif)





## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
